File: steps/canonical/ts/session.ts
import { readFileSync, writeFileSync, existsSync } from "fs";
import { join } from "path";

// The session is just the message array on disk. Save after every turn; load it
// back on --resume. No database — the whole conversation is already a plain array.
const SESSION_FILE = join(process.cwd(), ".mini-session.json");

//#region session
export function saveSession(messages: unknown[]): void {
  try { writeFileSync(SESSION_FILE, JSON.stringify(messages, null, 2)); } catch {}
}

export function loadSession(): unknown[] | null {
  if (!existsSync(SESSION_FILE)) return null;
  try { return JSON.parse(readFileSync(SESSION_FILE, "utf-8")); } catch { return null; }
}
//#endregion

File: steps/canonical/py/session.py
import json
import os

# The session is just the message list on disk. Save after every turn; load it
# back on --resume. No database — the whole conversation is already a plain list.
SESSION_FILE = os.path.join(os.getcwd(), ".mini-session.json")


#region session
def save_session(messages) -> None:
    try:
        with open(SESSION_FILE, "w", encoding="utf-8") as f:
            json.dump(messages, f, indent=2, default=lambda o: getattr(o, "model_dump", lambda: str(o))())
    except Exception:
        pass


def load_session():
    if not os.path.exists(SESSION_FILE):
        return None
    try:
        with open(SESSION_FILE, encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return None
#endregion
